common/thorhelper/thorread.cpp

```diff
@@ -25,33 +25,184 @@
+    const RtlRecord & sourceRecord = sourceMeta->queryRecordAccessor(true);
+    if (strsame(format, "csv"))
+    {
+        type_vals format = options->hasProp("ascii") ? type_string : type_utf8;
+        translator.setown(createRecordTranslatorViaCallback(projectedRecord, sourceRecord, format));
+    }
+    else if (strsame(format, "xml"))
+    {
+        translator.setown(createRecordTranslatorViaCallback(projectedRecord, sourceRecord, type_utf8));
+    }
+    else
+    {
+        if ((projectedMeta != sourceMeta) && (projectedCrc != sourceCrc))
+            translator.setown(createRecordTranslator(projectedRecord, sourceRecord));
+    }
+
+    if (translator)
+    {
+        DBGLOG("Record layout translator created for %s", filename);
+        translator->describe();
+
+        if (!translator->canTranslate())
+            throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s", filename);
+
+        if (translator->needsTranslate())
+        {
+            if (sourceMeta != expectedMeta)
+            {
+                Owned<const IKeyTranslator> _keyedTranslator = createKeyTranslator(sourceMeta->queryRecordAccessor(true), expectedMeta->queryRecordAccessor(true));
+                //MORE: What happens if the key filters cannot be translated?
+                if (_keyedTranslator->needsTranslate())
+                    keyedTranslator.swap(_keyedTranslator);
+            }
+        }
+        else
+            translator.clear();
+    }
+
+    checkedTranslators = true;
+}
+
+THORHELPER_API IDiskReadMapping * createDiskReadMapping(RecordTranslationMode mode, const char * format, unsigned actualCrc, IOutputMetaData & actual, unsigned expectedCrc, IOutputMetaData & expected, unsigned projectedCrc, IOutputMetaData & projected, const IPropertyTree * options)
+{
+    assertex(expectedCrc);
+    assertex(options);
+    return new DiskReadMapping(mode, format, actualCrc, actual, expectedCrc, expected, projectedCrc, projected, options);
+}
+
+THORHELPER_API IDiskReadMapping * createUnprojectedMapping(IDiskReadMapping * mapping)
+{
+    return createDiskReadMapping(mapping->queryTranslationMode(), mapping->queryFormat(), mapping->getActualCrc(), *mapping->queryActualMeta(), mapping->getExpectedCrc(), *mapping->queryExpectedMeta(), mapping->getExpectedCrc(), *mapping->queryExpectedMeta(), mapping->queryOptions());
+}
+
 
 //---------------------------------------------------------------------------------------------------------------------
 
 constexpr size32_t defaultReadBufferSize = 0x10000;
 
-class DiskRowReader : extends CInterfaceOf<IAllocRowStream>, implements IRawRowStream, implements IDiskRowReader, implements IThorDiskCallback
+/*
+ * The base class for reading rows from an external file.  Each activity will have an instance of a disk reader for
+ * each actual file format.
+ */
+class DiskRowReader : extends CInterfaceOf<IDiskRowStream>, implements IDiskRowReader, implements IThorDiskCallback
```

Comment:
style: similar to other q. - I have always thought 'extends' was something we introduced to be coupled with 'interface', in this context I'd use public .. ?